GeoJSON platform

exxamalte opened the pull request:
## Description:
This is the first platform of the new geo location component. It fetches data from a [GeoJSON](http://geojson.org/) feed and generates entities for each external event that is in the vicinity of the Home Assistant server. The radius around Home Assistant is configurable, and so is the scan interval.

I tried to keep this as simple as possible. Future enhancements may include the ability to access other properties of the feed or the ability to define additional filters.
More platforms for specific GeoJSON feeds are in the making, and so are similar platforms for GeoRSS feeds.

**Related issue (if applicable):** evolving https://github.com/home-assistant/home-assistant/pull/15953

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation (if applicable):** home-assistant/home-assistant.github.io#6316

## Example entry for `configuration.yaml` (if applicable):
```yaml
geo_location:
  - platform: geo_json_events
    url: http://www.rfs.nsw.gov.au/feeds/majorIncidents.json
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [x] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [x] New or updated dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [x] Tests have been added to verify that the new code works.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

exxamalte committed:
initial version of geojson platform

exxamalte committed:
unit tests for geo json platform added; smaller bugfixes and code cleanups

exxamalte committed:
fixing pylint issues

exxamalte commented:
@MartinHjelmare: Here comes the first platform for the new geo location component. I decided to go with GeoJSON feeds first before looking at GeoRSS again, because GeoJSON is a bit simpler and does not depend on the `feedreader` component.

MartinHjelmare reviewed (changes requested):
I didn't look at the tests.

Please rename device to entity consistently cause that's what is referred to here.

exxamalte commented on homeassistant/components/geo_location/geo_json_events.py:
```diff
@@ -0,0 +1,381 @@
+from homeassistant.components.sensor.rest import RestData
+from homeassistant.const import CONF_RADIUS, CONF_URL, CONF_SCAN_INTERVAL, \
+    EVENT_HOMEASSISTANT_START, ATTR_ID, LENGTH_KILOMETERS, LENGTH_METERS
+from homeassistant.helpers.config_validation import PLATFORM_SCHEMA
```
OK

exxamalte commented on homeassistant/components/geo_location/geo_json_events.py:
```diff
@@ -0,0 +1,381 @@
+    vol.Required(CONF_URL): cv.string,
+    vol.Optional(CONF_RADIUS, default=DEFAULT_RADIUS_IN_KM):
+        vol.Coerce(float),
+    vol.Optional(CONF_SCAN_INTERVAL, default=DEFAULT_SCAN_INTERVAL):
```
OK

exxamalte commented on homeassistant/components/geo_location/geo_json_events.py:
```diff
@@ -0,0 +1,381 @@
+def setup_platform(hass, config, add_entities, disc_info=None):
+    """Set up the GeoJSON Events platform."""
+    url = config.get(CONF_URL)
+    scan_interval = config.get(CONF_SCAN_INTERVAL)
```
OK

exxamalte commented on homeassistant/components/geo_location/geo_json_events.py:
```diff
@@ -0,0 +1,381 @@
+})
+
+
+def setup_platform(hass, config, add_entities, disc_info=None):
```
OK

exxamalte commented on homeassistant/components/geo_location/geo_json_events.py:
```diff
@@ -0,0 +1,381 @@
+    vol.Optional(CONF_SCAN_INTERVAL, default=DEFAULT_SCAN_INTERVAL):
+        cv.time_period,
+    vol.Optional(CONF_CATEGORIES, default=[]):
+        vol.All(cv.ensure_list, [cv.string]),
```
That won't be possible because the external GeoJSON feed can basically make up any category it wants.

MartinHjelmare replied:
Ok. How will the user know what to put here?

exxamalte replied:
The user would need to look at what categories the provider of the GeoJSON feed has documented, or do a bit of reverse-engineering.

MartinHjelmare replied:
Is there an example of categories from some established feed?

exxamalte replied:
Sure, for example this feed http://www.rfs.nsw.gov.au/feeds/majorIncidents.json uses the following categories:
* Emergency Warning
* Watch and Act
* Advice
* Not Applicable

Documented on the provider's [overview page](https://www.rfs.nsw.gov.au/fire-information/fires-near-me) when you scroll down to the table "Alert Level".

exxamalte replied:
I removed categories from this, and move this into a more specific platform, including validation.

exxamalte commented on homeassistant/components/geo_location/geo_json_events.py:
```diff
@@ -0,0 +1,381 @@
+                    device.longitude = longitude
+                    device.name = name
+                    device.category = category
+                    self._hass.add_job(device.async_update_ha_state())
```
OK.
Is there a corresponding function for removing an entity? I'm using `self._hass.add_job(entity.async_remove())` at the moment. Should I add the following to the `Entity` class?
```
def schedule_remove(self):
    self._hass.add_job(entity.async_remove())
```

MartinHjelmare replied:
No. Keep that as it is.

exxamalte commented on homeassistant/components/geo_location/geo_json_events.py:
```diff
@@ -0,0 +1,381 @@
+
+import homeassistant.helpers.config_validation as cv
+from homeassistant.components.geo_location import GeoLocationEvent
+from homeassistant.components.sensor.rest import RestData
```
OK, I understand the direction the project is taking here. 
However, at this point there would really not be much more than what is currently implemented in the `RestData` class that would need to be outsourced.
And before I go down that path of an external library, could you please have a look at the `GeoJsonDistanceHelper` inside the platform. Would you expect that to be outsourced, too, or does this still count as data transformation and can stay inside the platform?

MartinHjelmare replied:
`GeoJsonDistanceHelper` certainly looks general enough to fit in a standalone library. I think that's a good idea.

exxamalte replied:
Alright, let me take a look at moving the code.

exxamalte replied:
In the meantime I moved all code that interacts with the external GeoJSON feed as well as the code that calculates distances into a separate library: https://pypi.org/project/geojson-client/

exxamalte commented:
@MartinHjelmare : Sorry, I just realised from some of your questions that it may make sense to briefly discuss the general direction for new platforms in the geo location component. Could you please have a look at the last comment in here: https://github.com/home-assistant/architecture/issues/42

MartinHjelmare commented:
Those are good questions. I don't have an opinion/answer at the moment.

exxamalte commented on homeassistant/components/geo_location/geo_json_events.py:
```diff
@@ -0,0 +1,381 @@
+        """Return the name of the entity."""
+        return self._name
+
+    @name.setter
```
OK

exxamalte commented on homeassistant/components/geo_location/geo_json_events.py:
```diff
@@ -0,0 +1,381 @@
+        return DEFAULT_UNIT_OF_MEASUREMENT
+
+    @property
+    def external_id(self):
```
OK

exxamalte commented on homeassistant/components/geo_location/geo_json_events.py:
```diff
@@ -0,0 +1,381 @@
+        return self._external_id
+
+    @property
+    def category(self):
```
I removed the `category` altogether for this generic platform since there is no real standard defining categories in GeoJSON feeds.

exxamalte commented on homeassistant/components/geo_location/geo_json_events.py:
```diff
@@ -0,0 +1,381 @@
+                    # Update existing device's details with event data.
+                    latitude, longitude, _, name, category = \
+                        self._extract_data(feature)
+                    device.distance = entry[ATTR_DISTANCE]
```
OK. The manager class I implemented is currently fetching the data from the new library and can keep a copy of all entries. Each entity then fetches the original entry on `async_update` and updates its values.

exxamalte committed:
moved all code accessing the external feed into separate library; refactored platform and tests to work with that new library

exxamalte commented:
@MartinHjelmare : As suggested I moved all code that accesses the external GeoJSON feed into a separate library. I refactored the code in this PR as requested and to integrate with the new library. 
As suggested in the last comment in https://github.com/home-assistant/architecture/issues/42, this PR may not go anywhere, but before I go ahead and integrate a first specific GeoJSON feed as a new platform, I was hoping you could take a quick look at this to confirm whether the refactoring go into the right direction? Thanks.

exxamalte committed:
fixing lint

MartinHjelmare commented on homeassistant/components/geo_location/geo_json_events.py:
```diff
@@ -0,0 +1,198 @@
+        """No polling needed for GeoJSON location events."""
+        return False
+
+    @asyncio.coroutine
```
Use Python 3.5 async syntax, ie `async def`.

exxamalte replied:
OK

MartinHjelmare commented on homeassistant/components/geo_location/geo_json_events.py:
```diff
@@ -0,0 +1,198 @@
+
+def setup_platform(hass, config, add_entities, discovery_info=None):
+    """Set up the GeoJSON Events platform."""
+    url = config.get(CONF_URL)
```
Use `dict[key]` for required config keys.

exxamalte replied:
OK

MartinHjelmare reviewed (commented):
Looks ok.

exxamalte committed:
small refactorings

MartinHjelmare reviewed (approved):
Looks good!

exxamalte commented:
Updated the PR description. Doco will be added tonight.
